Added basic views

tunihoop opened the pull request:
Hey,
I added the basic views we need. No design.
Let me know your thoughts.
Thanks

tunihoop commented:
Hi,
I added model validation, please have a look and see if you agree.
Thanks

tunihoop commented:
Hi, 
Fixed small typos and bugs here.
Thanks

Stevinson commented on app/controllers/articles_controller.rb:
```diff
@@ -39,7 +41,7 @@ def destroy
   # Whitelist and return a hash of the params required
   def article_params
-    params.require(:article).permit(:title, :summary, :content)
+    params.require(:article).permit(:title, :summary, :url)
```
I think we need the price column to be whitelisted too?

tunihoop replied:
Yes, agree. Will add it

Stevinson commented on app/controllers/articles_controller.rb:
```diff
@@ -20,6 +20,8 @@ def create
 
   def show
+    @booking = Booking.new
+    @review = Review.new
```
Can you explain these to me tomorrow?

tunihoop replied:
Sure. Since the booking form and review form are on the article page, we need to pass an empty instance to fill in the form. Happy to show how it works tomorrow

Stevinson commented on app/controllers/bookings_controller.rb:
```diff
@@ -1,8 +1,9 @@
 class BookingsController < ApplicationController
   def index
     @bookings = Booking.all
+    @user = current_user
```
I tried to find the documentation but failed - what does 'current_user' return?

tunihoop replied:
The active/logged-in user

Stevinson commented on app/models/article.rb:
```diff
@@ -2,4 +2,7 @@ class Article < ApplicationRecord
   has_many :reviews
+  validates :title, presence: true, length: { in: 3..150 }
+  validates :summary, presence: true, length: { in: 50..400 }
+  validates :url, presence: true, allow_blank: false
```
We could add a regex to make sure it's in the format of a url?

Stevinson replied:
We should also provide a validation that the price is numerical

tunihoop replied:
I found a regex and rails didn't like it. Will try to find a simpler one.
Agree on price, will add

Stevinson commented on app/models/booking.rb:
```diff
@@ -2,4 +2,14 @@ class Booking < ApplicationRecord
+
+  def start_must_be_before_end_date
+    errors.add(:start_date, "must be before end time") unless
+        start_date < end_date
```
How does the variable 'end_date' have access to the end date?

tunihoop replied:
The validation just looks at conditions and see whether they're checked. So here we compare start and end dates. I am adding a validation for start_date > Date.today

Stevinson commented on app/models/booking.rb:
```diff
@@ -2,4 +2,14 @@ class Booking < ApplicationRecord
+  validates :start_date, presence: true
+  validates :end_date, presence: true
+  validate :start_must_be_before_end_date
+
```
Shall we make it so that the user/article/time combo so that a user cannot double-purchase an article?

tunihoop replied:
This is done on the article view page. The booking form is only displayed if the article is not currently accessible

Stevinson commented on app/controllers/bookings_controller.rb:
```diff
@@ -1,8 +1,9 @@
     @bookings = Booking.all
+    @user = current_user
   end
 
```
Does the show method need access to a private 'calculate_price' method, so we can show the cost of the booking

tunihoop replied:
The booking price is set as a parameter when we create the booking (as it is a required attribute). We actually calculate it on the booking form (we need to so that a user can see how the price changes when selecting dates), and pass it as a param

Stevinson commented on app/models/user.rb:
```diff
@@ -1,9 +1,17 @@
          :recoverable, :rememberable, :trackable, :validatable
+
+  def today_booked_articles
+    booked_articles.where("bookings.start_date <= :date AND bookings.end_date >= :date", date: Date.today)
```
This logic is as a string? Explain tomorrow?

tunihoop replied:
This method renders an array of articles from the db, which are under current booking by the user. As you mentioned and explained above, we do not want a user to double-book an article, so we check on the show page whether it's currently amongst the articles booked. Happy to further explain tomorrow

Stevinson commented on app/views/articles/_form.html.erb:
```diff
@@ -0,0 +1,19 @@
+<div class="container">
+  <div class="row">
+    <div class="col-xs-12">
```
Let's make it adapt to other screens (and have an offset)

tunihoop replied:
Will address formatting tomorrow

Stevinson commented on app/views/articles/edit.html.erb:
```diff
@@ -1,2 +1 @@
-<h1>Articles#edit</h1>
-<p>Find me in app/views/articles/edit.html.erb</p>
+<%= render 'form' %>
```
Let's put in a container

tunihoop replied:
Will address formatting tomorrow

tunihoop commented:
Hi,
I added update and delete options to articles -only for owners.
I cleaned forms. However booking form not functional. The rest works.
Let me know if you manage to fix the booking form.
Thanks

Stevinson commented on app/views/articles/index.html.erb:
```diff
@@ -1,2 +1,47 @@
+<div class="container">
+
+  <div class="row">
+    <div class="col-xs-12">
```
Adaptive

Stevinson replied:
Also, the title shouldn't really be in a row

Stevinson commented on app/views/articles/index.html.erb:
```diff
@@ -1,2 +1,47 @@
+            <% elsif (i-1..i).include? average_rating  %>
+              <i class="fa fa-star-half-o" aria-hidden="true"></i>
+            <% else %>
+              <i class="fa fa-star-o" aria-hidden="true"></i>
```
I understand most of this (it looks cool)

Stevinson commented on app/views/articles/index.html.erb:
```diff
@@ -1,2 +1,47 @@
+      <% end %>
+
+    <% end %>
+    </div>
```
We should prob give some of the elements in the cards classes so we can add CSS

tunihoop replied:
Yes! We need to draw a card together and agree which elements to place and where so that we can reshuffle in html if needed and agree on css

Stevinson commented on app/views/articles/new.html.erb:
```diff
@@ -1,2 +1 @@
-<h1>Articles#new</h1>
-<p>Find me in app/views/articles/new.html.erb</p>
+<%= render 'form' %>
```
Put in a div

Stevinson commented on app/views/articles/show.html.erb:
```diff
@@ -1,2 +1,77 @@
+
+      <!-- Show URL/content only if accessible, i.e. owned or booked -->
+
+      <% accessible_articles = current_user.articles + current_user.today_booked_articles %>
```
1. I don't think you have access to today_booked_articles
2. I have no idea what 'today_booked_articles' is, could you explain?

tunihoop replied:
The articles accessible to the user at the moment, i.e. the ones they booked, with start_date <= today and end_date >= today

tunihoop replied:
(the equivalent of current trip in airbnb)

Stevinson commented on app/views/articles/show.html.erb:
```diff
@@ -1,2 +1,77 @@
+        <%= link_to @article.url, @article.url %>
+      <% else %>
+        <p>You do not have access to this article</p>
+      <% end %>
```
What is this doing?

tunihoop replied:
Displays URL (or whatever is purchased through the booking) if it is currently booked. Otherwise we notify the user they do not have access, and they will have on the RHS a CTA (booking form)

Stevinson commented on app/views/articles/show.html.erb:
```diff
@@ -1,2 +1,77 @@
+    </div>
+  </div>
+
+
```
I'll have to look at this rendered, because it's baffling me a little

Stevinson replied:
I only got up to here - I'm knackered. I'll look at the rest tomorrow morning

tunihoop replied:
Sure - the booking form doesn't work so will need to fix it as well. We can discuss tomorrow

tunihoop commented:
As agreed, I added url and price validations to article model. Added price param to articles controller.
Please note the url validation requires a gem which I add, do not forget the bundle install.
I am knackered too. Will tackle the formatting aspects tomorrow.
Cheers
